source/MRMesh/MRMesh.cpp

```diff
@@ -521,41 +522,157 @@ float Mesh::discreteMeanCurvature( UndirectedEdgeId ue ) const
+    }
 
     UndirectedEdgeBitSet takeEdges() { return std::move( edges_ ); }
+    ConnectionsPerVert takeSelectedVertsPerVert() { return std::move( connectionsPerVert_ ); }
 
     void operator()( const tbb::blocked_range<UndirectedEdgeId> & r ) 
     {
         for ( UndirectedEdgeId ue = r.begin(); ue < r.end(); ++ue ) 
         {
             if ( mesh_.topology.isLoneEdge( ue ) )
                 continue;
+
             auto dihedralCos = mesh_.dihedralAngleCos( ue );
+
             if ( dihedralCos <= critCos_ )
+            {
                 edges_.set( ue );
+                const VertId vOrg = mesh_.topology.org( ue );
+                const VertId vDest = mesh_.topology.dest( ue );
+                connectionsPerVert_[vOrg].push_back( { ue, vDest } );
+                connectionsPerVert_[vDest].push_back( { ue, vOrg } );
+            }
         }
     }
 
 private:
     const Mesh & mesh_;
     float critCos_ = 1;
+    float critLength_ = 0;
+    bool filterBranches_ = false;
     UndirectedEdgeBitSet edges_;
+    ConnectionsPerVert connectionsPerVert_;
 };
 
-UndirectedEdgeBitSet Mesh::findCreaseEdges( float angleFromPlanar ) const
+UndirectedEdgeBitSet Mesh::findCreaseEdges( float angleFromPlanar, float critLength, bool filterBranches ) const
 {
     MR_TIMER
     assert( angleFromPlanar > 0 && angleFromPlanar < PI );
     const float critCos = std::cos( angleFromPlanar );
-    CreaseEdgesCalc calc( *this, critCos );
+    CreaseEdgesCalc calc( *this, critCos, critLength, filterBranches );
     parallel_reduce( tbb::blocked_range<UndirectedEdgeId>( 0_ue, UndirectedEdgeId{ topology.undirectedEdgeSize() } ), calc );
-    return calc.takeEdges();
+
+    auto selectedEdges = calc.takeEdges();
+    auto selectedComponents = MeshComponents::getAllComponentsUndirectedEdges( *this, selectedEdges );
+
+    for ( const auto& selectedComponent : selectedComponents )
+    {
+        const auto componentLength = parallel_deterministic_reduce( tbb::blocked_range( 0_ue, UndirectedEdgeId{ selectedComponent.size() }, 1024 ), 0.0,
+       [&] ( const auto& range, double curr )
+            {
+                for ( UndirectedEdgeId ue = range.begin(); ue < range.end(); ++ue )
+                    if ( selectedComponent.test( ue ) )
+                        curr += edgeLength( ue );
+                return curr;
+            },
+       [] ( auto a, auto b )
+        {
+            return a + b;
+        } );
+
+        if ( componentLength < critLength )
+        {
+            BitSetParallelFor( selectedComponent, [&] ( UndirectedEdgeId ue )
+            {
+                selectedEdges.set( ue, false );
+            } );
+        }
+    }
+    
+    if ( !filterBranches )
+        return selectedEdges;
+    
+    auto connectionsPerVert = calc.takeSelectedVertsPerVert();
+
+    for ( VertId i = VertId( 0 ); i < connectionsPerVert.size(); ++i )
+    {
+        auto& connections = connectionsPerVert[i];
+        if ( connections.size() != 1 )
```

Comment:
instead of calculating `connectionsPerVert` its better to
```c++
auto incidentVertices = getIncidentVerts( topology,  selectedEdges);
// ... 
int numConnections = 0;
for ( auto e : orgRing( topolgy, v ) )
    if ( selectedEdges.test( e.undirected() ) )
        ++numConections;
// then track it down until first vertex with numConnections == 1 || numConnections > 2

```